---
title: "carlos_simon_RNAseq"
author: "Sara"
date: "2025-09-05"
output: html_document
---

```{r}
library(dplyr)
library (DESeq2)
library(ggplot2)
library(EnhancedVolcano)
library(clusterProfiler)
library(org.Hs.eg.db)
library(pheatmap)
library(RColorBrewer)

```

#Load data

#Individual counts were generated using subread funcition "feature count", load the files and merge them into a count matrix with geneids as row names and samples as column names

```{r}
#Load all inidivual counts and merge in to a count matrix
setwd("~/cv/BIOINFORMATICS_VALENCIA/technical_test/feature_counts")
control_1 = read.table("control_1_genome.txt", sep="\t", header =TRUE)
control_2 = read.table("control_2_genome.txt", sep="\t", header =TRUE)
control_3 = read.table("control_3_genome.txt", sep="\t", header =TRUE)
treated_1 = read.table("treated_1_genome.txt", sep="\t", header =TRUE)
treated_2 = read.table("treated_2_genome.txt", sep="\t", header =TRUE)
treated_3 = read.table("treated_3_genome.txt", sep="\t", header =TRUE)


gsub("\\..*","",a)


counts  <- list (control_1, control_2, control_3, treated_1, treated_2, treated_3)

#Remove all extra columns generated by subread(lenght, chromosome, etc) 
#Select only column 1 (Geneid) and column 7 (gene counts)
counts <- lapply(counts, function(df) df[, c(1, 7)])

#Merge all dataframes together
count_matrix <- Reduce(function(x, y) merge(x, y, by = "Geneid", all = TRUE), counts)

#Sumarize reads by adding up all different transcript versions - for this case we are about information at the gene level, not trasncript
count_matrix <-  count_matrix %>%
    mutate(Geneid = sub("\\..*", "", Geneid)) %>%  # remove everything after '.', keeps the ENBML id for the gene rather than the transcript
    group_by(Geneid) %>%
    summarise(across(everything(), sum), .groups = "drop")


#Set Geneid as row names and remove it as a column 
count_matrix <- as.data.frame(count_matrix)
row.names(count_matrix) <- as.factor(count_matrix$Geneid)
count_matrix <- count_matrix[ -1]

#Load also the sample information
sample_info = read.csv("sample_info.csv", row.names=1)

```

##Create deseq2 object 
```{r}
colnames(count_matrix)
row.names(sample_info)

dds <- DESeqDataSetFromMatrix(countData= count_matrix, 
                              colData=sample_info, design= ~Treatment)
dds 
```

##Run Deseq2
```{r}
#Remove any transcripts that have 0 across all samples
notAllZero <- (rowSums(counts(dds))>0) 
dds <- dds[notAllZero,]
dds

#Run deseq2,  
dds$Treatment <- relevel( dds$Treatment, "control" )
as.data.frame(colData(dds) )
dds <- DESeq(dds)
rld <- rlog(dds, blind=FALSE)
```

#PCA - sample distribution

```{r}
#PCA
object <- rld
ntop <- 500
intgroup <- c("Treatment")
#PCA breakdown
# calculate the variance for each gene
rv <- rowVars(assay(object))
# select the ntop genes by variance
select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]

# perform a PCA on the data in assay(x) for the selected genes
pca <- prcomp(t(assay(object)[select,]))

# the contribution to the total variance for each component
percentVar <- as.data.frame(pca$sdev^2 / sum( pca$sdev^2 ))
if (!all(intgroup %in% names(colData(object)))) {
  stop("the argument 'intgroup' should specify columns of colData(dds)")
}
intgroup.df <- as.data.frame(colData(object)[, intgroup, drop=FALSE])

# add the intgroup factors together to create a new grouping factor
group <- if (length(intgroup) > 1) {
  factor(apply( intgroup.df, 1, paste, collapse=":"))
} else {
  colData(object)[[intgroup]]
}
# assembly the data for the plot
pcaData <- data.frame(PC1=pca$x[,1], PC2=pca$x[,2], group=group, intgroup.df, name=colnames(object))


#Plot using ggplot
x_axis <-   paste0("PC1: ",round(percentVar[1, ] * 100),"% variance")
y_axis <- paste0("PC2: ",round(percentVar[2, ] * 100),"% variance")


pca_plot <- ggplot(pcaData, aes(x = PC1, y = PC2, color = factor(Treatment),
                    shape = factor(Treatment))) +
  geom_point(size =5)+
  scale_shape_manual (values = c(19, 17))+
  scale_fill_manual(values = c("control" = "#D55E00", "treated" = "#56B4E9"))+
  theme(text = element_text(size = 17)) + 
  xlab(x_axis) + ylab (y_axis)
  theme_light()
  
  
  
pca_plot

#Save PCA plot as pdf 
setwd("~/cv/BIOINFORMATICS_VALENCIA/technical_test/")
pdf("pca_plot.pdf", height=5, width=8)
pca_plot
dev.off
  
```



#Checkout results from deseq2
```{r fig.height=9 , fig.width=9}
cont_resuls <- results(dds, contrast = c("Treatment", "treated", "control"),
                       lfcThreshold=0)

cont_resuls

#Add gene symbols to the results for easy recognistion 
symbols <- mapIds(org.Hs.eg.db, keys = rownames(cont_resuls),
                  column = c('SYMBOL'), keytype = 'ENSEMBL')
cont_resuls$symbol <- symbols[rownames(cont_resuls)]
head(cont_resuls)

#Count all up and down genes
upregulated_genes <- sum(cont_resuls$log2FoldChange > 1.5 & cont_resuls$padj < 0.05 & complete.cases(cont_resuls))
downregulated_genes <- sum(cont_resuls$log2FoldChange < -1.5 & cont_resuls$padj < 0.05 & complete.cases(cont_resuls))

# Order the results by ascending p-value and select top 20 genes to label on the volcano plot 
cont_resuls_ordered <- cont_resuls[order(cont_resuls$padj), ]
top_genes <- head(cont_resuls_ordered$symbol, 20)
top_genes

#Crete the volcano plot
volcano <- EnhancedVolcano(cont_resuls,
    lab = cont_resuls$symbol,
    x = 'log2FoldChange',
    y = 'padj',
    title = "Differentially expressed genes",
    selectLab = c(top_genes),
    boxedLabels = TRUE,
    drawConnectors = TRUE,
    xlim = c(-7,7),
    ylim = c(0,100),
    FCcutoff = 1.5,
    pCutoff = 0.05,
    pointSize = 2.5,
    labSize = 3.5,
    legendLabSize = 20,
    widthConnectors = 0.0055,
    colConnectors = 'black',
    col=c('black', 'black', 'black', "cyan3")
    )+
  annotate("text", x = +6.5, y = 97, label = paste("Up:", upregulated_genes),
           hjust = 1, vjust = 0, size = 8) +
  annotate("text", x = -6.5, y = 97, label = paste("Down:", downregulated_genes),
           hjust = 0, vjust = 0, size = 8)

#PLOT
volcano


#Save plot as pdf 
setwd("~/cv/BIOINFORMATICS_VALENCIA/technical_test/")
pdf("volcanoplot.pdf", height=12, width=12)
volcano
dev.off
```

```{r}
#Make a heatmap of all DEGS 
counts_norm <- as.data.frame((assay(rld)))
counts_norm[ ,"gene"]<- row.names(counts_norm)

DEGs <- c(rownames(up_genes), rownames(down_genes))

significant_subset <- subset(counts_norm, counts_norm$gene %in% DEGs)
significant_subset <- subset(significant_subset, select = -c(gene)) 

significant_subset_scaled <- t(scale(t(significant_subset)))

heatmap <- pheatmap(significant_subset_scaled,
         show_rownames = FALSE, 
         fontsize_row = 8,
         col = colorRampPalette(brewer.pal(9, "YlOrRd"))(100),
         fontsize_col=9, main = "All DEGs ")
```


##Biological processed affected (GO enrichment)
```{r fig.height=4, fig.width=5}
#GO enrichment on upregulated DEGs
up_genes <- cont_resuls[cont_resuls$log2FoldChange > 1.5 & cont_resuls$padj < 0.05 & complete.cases(cont_resuls), ]                
                                                                          
up_entrez = bitr(rownames(up_genes), fromType = "ENSEMBL", 
                       toType = c("ENTREZID"),OrgDb = org.Hs.eg.db)  

up_enrich = enrichGO(gene = up_entrez$ENTREZID, 
                      OrgDb = org.Hs.eg.db,readable = T,ont = "BP",
                      pvalueCutoff = 0.05, qvalueCutoff = 0.10)

barplot(up_enrich, showCategory=5) 
```

```{r}
#GO enrichment on upregulated DEGs
down_genes <- cont_resuls[cont_resuls$log2FoldChange < -1.5 & cont_resuls$padj < 0.05 & complete.cases(cont_resuls), ]                
                                                                          
down_entrez = bitr(rownames(down_genes), fromType = "ENSEMBL", 
                       toType = c("ENTREZID"),OrgDb = org.Hs.eg.db)  

down_enrich = enrichGO(gene = down_entrez$ENTREZID, 
                      OrgDb = org.Hs.eg.db,readable = T,ont = "BP",
                      pvalueCutoff = 0.05, qvalueCutoff = 0.10)

barplot(down_enrich, showCategory=10) # no significant result 

```

#Checkout expression of specific genes
```{r}
#Extract normalized (but not log transformed) counts from deseq object.
expression <- counts(dds, normalized=TRUE)
#Change rownames to gene symbols for easy gene calling
rownames(expression)<- symbols
#Merge expression with sample info for plotting
sample_info <- t(sample_info)
expression <- rbind (expression, sample_info)
```

```{r}
#Create function to plot barplot of normalized reads for one gene
plot_gene_error_bars <- function(gene) {
  
  #Subset expression dataframe to only contain expression of gene of interest and Treatment information 
  select <- c(gene, "Treatment") 
  gene_only <- as.data.frame(t(expression[select, ]))
  gene_only[ ,1] <- as.numeric(  gene_only[ ,1]) 
  gene_only$Treatment <- as.factor(gene_only$Treatment)
  
  #Calculate the mean and standard deviation
    gene_mean <- gene_only %>% 
    group_by(Treatment, .groups = 'drop') %>%
    summarise(across(gene, list(mean = ~ mean(.x),sd = ~ sd(.x)   
    )))
  colnames(gene_mean) <- c("Treatment", ".groups","mean", "sd")
  
  #Peform a t-test to check if there is a significant difference
  t_test_result <- t.test(gene_only[[1]] ~ gene_only$Treatment)
  pval <- t_test_result$p.value
  
  #Decide values of significance value in the graph 
  sig_label <- ifelse(pval < 0.05, "*", "ns") 
  y_max <- max(gene_mean$mean + gene_mean$sd) 
  y_line <- y_max * 1.1 #y position 1.1 points above 
  
  #Plot 
  p <- ggplot(gene_mean, aes(x = Treatment, y = mean, fill = Treatment)) +
    geom_bar(stat = "identity", position = "dodge") +
    geom_errorbar(aes(ymin = mean - sd, ymax = mean + sd), position = position_dodge(width = 0.9), width = 0.25) +
    geom_jitter(data = gene_only, aes(x = Treatment, y = gene_only[ ,1]),
                width = 0.15, size = 2, color = "black") +
    labs(x = "Treatment", y = "Normalized counts", fill = "Treatment", title = gene) +
     geom_segment(aes(x = 1, xend = 2, y = y_line, yend = y_line)) +
    geom_text(aes(x = 1.5, y = y_line * 1.05, label = sig_label), size = 5)+
    theme_minimal() +
    scale_fill_manual(values = c("control" = "#D55E00", "treated" = "#56B4E9"))
  return(p)

  
}
```

```{r fig.width=3.5, fig.height=4}
plot_gene_error_bars("FOXP3")
plot_gene_error_bars("LIF")

#Save plot as pdf 
setwd("~/cv/BIOINFORMATICS_VALENCIA/technical_test/")
pdf("FOXP3_expression.pdf", height=4, width=3.5)
plot_gene_error_bars("FOXP3")
dev.off
```

